### PR TITLE
Support MTP + CP

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

Enable context parallelism for DeepSeek-V3 MTP, including combined CP+TP with
sequence parallelism. MTP inputs are shifted before CP sharding so tokens,
labels, and validity masks remain aligned across rank and packed-document
boundaries.

## Design

- **Shift, then shard.** Build MTP tokens, labels, and masks from the complete
  per-DP-rank sequence, then apply the same CP load balancing and sharding to
  all aligned tensors. This avoids a separate cross-rank halo exchange.
- **Respect packed documents.** Accept a shifted value only when
  `positions[i + shift] == positions[i] + shift`; otherwise mask the input and
  set its label to `IGNORE_INDEX`.
- **Align masks under SP.** Redistribute the 1D validity mask from the decoder
  token-ID placement (TP-replicated) to the sequence-parallel token-ID
  placement (TP-sharded along the token dimension).
- **Preserve SPMD loss types.** Initialize the auxiliary sum from a real CE
  result and annotate `global_valid_tokens` before final normalization.

## Tests

Targeted CPU tests:

```bash
pytest -q \
  tests/unit_tests/cpu/test_context_parallel_validation.py \
  tests/unit_tests/cpu/test_loss.py
```

Result: `35 passed`.

## Numerical validation

The CP-off baseline (`4a797f3d6`) and CP-on change (`582637a15`) were
compared for 10 deterministic training steps from the same seed checkpoint.
Both used `spmd_types`, seed 42, a 256-token context, 256 tokens per training
step, and disabled CUDA graphs. Results use full-precision TensorBoard values.

| Configuration | GPUs (baseline/change) | Maximum loss relative difference | Maximum grad-norm relative difference |
| --- | ---: | ---: | ---: |
| TP=1, CP=1 -> CP=2 | 1 / 2 | `7.20e-5` | `7.54e-4` |
| TP=2 with SP, CP=1 -> CP=2 | 2 / 4 | `2.63e-4` | `3.79e-4` |

Both comparisons were repeated. Every configuration reproduced exactly across
its two runs.

All runs completed successfully and followed the same loss and gradient-norm
trajectories. Bitwise equality is not expected across CP=1 and CP=2 because CP
changes attention and collective reduction ordering.

### CODE REVIEW

File: tests/integration_tests/models.py [hunk 1]
```diff
@@ -54,6 +54,18 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
             # H100/GB200; skip on ROCm where they are unvalidated.
             skip_rocm_test=True,
         ),
+        OverrideDefinitions(
+            configs=[recipes.deepseek_v3_debugmodel_mtp_cp2],
+            test_descr="DeepSeek V3 MTP CP",
+            test_name="deepseek_v3_mtp_cp",
+            ngpu=2,
+        ),
+        OverrideDefinitions(
```
Comment (tianyu-l) [inline tests/integration_tests/models.py:57]: adding this one feels enough

File: torchtitan/models/deepseek_v3/mtp.py [hunk 2]
```diff
@@ -425,11 +422,17 @@ def __call__(
             )
         mtp_weight = self.mtp_scale / num_mtp_layers
         main_loss, _ = super().__call__(pred[0], labels[0])
-        mtp_loss = pred[0].new_zeros((), dtype=torch.float32)
-        for mtp_pred, mtp_labels in zip(pred[1:], labels[1:], strict=True):
+        mtp_loss, _ = super().__call__(pred[1], labels[1])
+        mtp_loss = mtp_loss * mtp_weight
+        for mtp_pred, mtp_labels in zip(pred[2:], labels[2:], strict=True):
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: what's the motivation of this change?

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/deepseek_v3/mtp.py (hunk 2 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py]: asked the same question to agent lol

With CP added, now mtp_loss should be partial on DP and CP, however if  we created `mtp_loss = pred[0].new_zeros((), dtype=torch.float32)`, where pred[0] is sharded on DP and CP, `mtp_loss` will become replicated. 

Before we are not turning on spmd type checking.

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/mtp.py]: use mutate_type API from spmd_types
